src/java.rs

```diff
@@ -0,0 +1,224 @@
+use crate::mapper::ProguardMapper;
+
+fn java_base_types(encoded_ty: char) -> Option<&'static str> {
+    match encoded_ty {
+        'Z' => Some("boolean"),
+        'B' => Some("byte"),
+        'C' => Some("char"),
+        'S' => Some("short"),
+        'I' => Some("int"),
+        'J' => Some("long"),
+        'F' => Some("float"),
+        'D' => Some("double"),
+        'V' => Some("void"),
+        _ => None,
+    }
+}
+
+fn byte_code_type_to_java_type(byte_code_type: &str, mapper: &ProguardMapper) -> Option<String> {
+    let mut chrs = byte_code_type.chars();
+    let token = chrs.next()?;
+    if token == 'L' {
+        // invalid signature
+        let l = chrs.clone().last();
+        if l.is_none() || l.unwrap() != ';' {
+            return None;
+        }
+        chrs.next_back(); // remove final `;`
+        let obfuscated = chrs.collect::<String>().replace('/', ".");
+
+        if let Some(mapped) = mapper.remap_class(&obfuscated) {
+            return Some(mapped.to_string());
+        }
+
+        return Some(obfuscated);
+    } else if token == '[' {
+        let type_sig = chrs.clone().collect::<String>();
+        if !type_sig.is_empty() {
+            return Some(format!(
+                "{}[]",
+                byte_code_type_to_java_type(chrs.collect::<String>().as_str(), mapper)
+                    .unwrap_or_default()
+            ));
+        }
+    } else if let Some(ty) = java_base_types(token) {
+        return Some(ty.to_string());
+    }
+    Some(byte_code_type.to_string())
+}
+
+// parse_obfuscated_bytecode_signature will parse an obfuscated signatures into parameter
+// and return types that can be then deobfuscated
+fn parse_obfuscated_bytecode_signature(signature: &str) -> Option<(Vec<String>, String)> {
+    let mut chrs = signature.chars();
+
+    let token = chrs.next();
+    if token.unwrap_or_default() != '(' {
+        return None;
+    }
+
+    let sig = chrs.collect::<String>();
+    let split_sign = sig.rsplitn(2, ')').collect::<Vec<&str>>();
+    if split_sign.len() != 2 {
+        return None;
+    }
+
+    let return_type = split_sign[0];
+    let parameter_types = split_sign[1];
```

Comment:
```suggestion
    let (parameter_types, return_type) = signature.rsplit_once(')')?;
```

I believe you had the variable names wrong here. The formatting should be `($parameter_types) $return_type`, right?

The `rsplit_once` solution looks a lot simpler and avoids an allocation.
It does however suffer from the same problem as the previous code that it does not support multiple `)` in the string. Although I’m not quite sure if that is legal to begin with?